app/routes/user/endorsements.server.ts

```diff
@@ -0,0 +1,214 @@
+    this.validateStateAndExistanceOfRequest(existingEndorsement)
+
+    await db.circular_endorsements.update({
+      Key: {
+        requestorSub: requestorSub,
+        endorserSub: this.#sub,
+      },
+      UpdateExpression: 'set #status = :status',
+      ExpressionAttributeNames: {
+        '#status': 'status',
+      },
+      ExpressionAttributeValues: {
+        ':status': status,
+      },
+    })
+  }
+
+  // Deletes the endorsement request the current user has requested from the user with a sub corresponding to endorserSub
+  async deleteEndorsementRequest(endorserSub: string) {
+    const db = await tables()
+    await db.circular_endorsements.delete({
+      requestorSub: this.#sub,
+      endorserSub,
+    })
+  }
+
+  /**
+   * Gets all of the current endorsements requests for a given user based on role
+   *
+   * 'requestor' returns all requests which the current user has made, 'endorser'
+   * returns all requests from other users in which the current user is the listed Endorser
+   *  */
+  async getEndorsements(
+    role: 'endorser' | 'requestor'
+  ): Promise<EndorsementRequest[]> {
+    const queryParams = {
+      IndexName:
+        role == 'requestor' ? undefined : 'circularEndorsementsByEndorserSub',
+      KeyConditionExpression:
+        role == 'requestor'
+          ? '#requestorSub = :sub'
+          : '#endorserSub = :endorserSub',
+      FilterExpression: role == 'requestor' ? undefined : '#status = :status',
+      ExpressionAttributeNames: {
+        '#requestorSub': 'requestorSub',
+        '#requestorEmail': 'requestorEmail',
+        '#endorserSub': 'endorserSub',
+        '#endorserEmail': 'endorserEmail',
+        '#status': 'status',
+        '#created': 'created',
+      },
+      ExpressionAttributeValues:
+        role == 'requestor'
+          ? {
+              ':sub': this.#sub,
+            }
+          : {
+              ':endorserSub': this.#sub,
+              ':status': 'pending',
+            },
+      ProjectionExpression:
+        '#requestorSub, #requestorEmail, #endorserSub, #endorserEmail, #status, #created',
+    }
+
+    const db = await tables()
+    const result = await db.circular_endorsements.query(queryParams)
+    return result.Items as EndorsementRequest[]
+  }
+
+  // Checks if the given sub corresponds to a user in the submitters group
+  async requestedEndorserIsValid(endorserSub: string) {
+    if (endorserSub == this.#sub)
+      throw new Response(
+        'Users can not request themselves as the endorser of their own request',
+        {
+          statusText: 'Bad Request',
+          status: 400,
+        }
+      )
+    const validUsers = await getSubmitterUsersFromCognito()
+    if (!validUsers)
+      throw new Response('No submitter users exist yet', {
+        statusText: 'Bad Request',
+        status: 400,
+      })
+    return validUsers?.some(
+      (user) =>
+        user.Attributes?.find((attr) => attr.Name == 'sub')?.Value ==
+        endorserSub
+    )
```

Comment:
Use Cognito's [ListUsers](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ListUsers.html) API with a Filter to find the user with the matching sub, then check that they belong to the submitters group. That way you don't get the entire Cognito database back!

Reply:
I don't think thats quite right for this case, the ListUsers method doesnt return info about the groups even when you pass the filter, just an object like the following:
```
{
  Attributes: [
    { Name: 'sub', Value: 'sub-string' },
    {
      Name: 'identities',
      Value: '[{"userId":"stuff","providerName":"stuff","providerType":"type","issuer":null,"primary":true,"dateCreated":DateTime}]'
    },
    { Name: 'email_verified', Value: 'false' },
    { Name: 'email', Value: 'example@example.com' }
  ],
  Enabled: true,
  MFAOptions: undefined,
  UserCreateDate: DateTime,
  UserLastModifiedDate: DateTime,
  UserStatus: 'EXTERNAL_PROVIDER',
  Username: 'usernamestring'
}
```
getSubmitterUsersFromCognito is just using the ListUsersInGroup. 

The docs also say:
You can only search for the following standard attributes:

username (case-sensitive)

email

phone_number

name

given_name

family_name

preferred_username

cognito:user_status (called Status in the Console) (case-insensitive)

status (called Enabled in the Console) (case-sensitive)

sub

Reply:
Right. So use ListUsers to find the user by sub, then AdminListGroupsForUser to get the user's groups. That way you avoid ever having a huge API response that contains the entire Cognito user database.

Reply:
Ahhh okay that makes sense. Going to implement it now

Reply:
Also, remember to add those actions to the arc plugin that grants the Lambda the IAM permissions to execute those API calls.